File: projects/lda_iris_segmentation/python/iris_lda.py
# lda_iris.py
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

def main():
    # Load data
    iris = load_iris()
    X = iris.data
    y = iris.target
    target_names = iris.target_names

    # Train/test split
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.25, random_state=42, stratify=y
    )

    # Fit LDA (classifier)
    lda = LinearDiscriminantAnalysis()
    lda.fit(X_train, y_train)

    # Predict
    y_pred = lda.predict(X_test)

    # Evaluate
    print("Accuracy:", accuracy_score(y_test, y_pred))
    print("\nConfusion matrix:\n", confusion_matrix(y_test, y_pred))
    print("\nClassification report:\n", classification_report(y_test, y_pred, target_names=target_names))

    # Optional: project to LDA space (max 2 components for 3 classes -> 2D)
    try:
        import matplotlib.pyplot as plt

        X_lda = lda.transform(X)  # shape: (n_samples, 2)
        plt.figure()
        for cls in sorted(set(y)):
            plt.scatter(X_lda[y == cls, 0], X_lda[y == cls, 1], label=target_names[cls])
        plt.title("Iris - LDA projection (2D)")
        plt.xlabel("LD1")
        plt.ylabel("LD2")
        plt.legend()
        plt.tight_layout()
        plt.show()
    except ImportError:
        print("\nmatplotlib not installed; skipping plot.")

if __name__ == "__main__":
    main()

File: projects/lda_iris_segmentation/R/iris_lda.R
# lda_iris.R

# Load required library
library(MASS)

# Load dataset (built into R)
data(iris)

# Set seed for reproducibility
set.seed(42)

# Train/test split (75% train, 25% test)
n <- nrow(iris)
train_index <- sample(1:n, size = 0.75 * n)

train_data <- iris[train_index, ]
test_data  <- iris[-train_index, ]

# Fit LDA model
lda_model <- lda(Species ~ ., data = train_data)

# Print model summary
print(lda_model)

# Predict on test set
pred <- predict(lda_model, test_data)

# Predicted classes
predicted_classes <- pred$class

# Confusion matrix
conf_matrix <- table(Predicted = predicted_classes,
                     Actual = test_data$Species)
print(conf_matrix)

# Accuracy
accuracy <- mean(predicted_classes == test_data$Species)
cat("\nAccuracy:", accuracy, "\n")

# Optional: Plot LDA projection (first two linear discriminants)
plot(pred$x[, 1], pred$x[, 2],
     col = as.numeric(test_data$Species),
     pch = 19,
     xlab = "LD1",
     ylab = "LD2",
     main = "Iris - LDA Projection")
legend("topright",
       legend = levels(iris$Species),
       col = 1:length(levels(iris$Species)),
       pch = 19)

